src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -345,9 +346,21 @@ class SelectableEntity(Selectable):
     Mainly used by session.table().
     """
 
-    def __init__(self, entity_name: str, *, analyzer: "Analyzer") -> None:
+    def __init__(
+        self,
+        entity_name: str,
+        *,
+        analyzer: "Analyzer",
+        is_generated_temp_table: bool = False,
+    ) -> None:
         super().__init__(analyzer)
         self.entity_name = entity_name
+        if is_generated_temp_table:
+            analyzer.session._inc_temp_table_ref_count(entity_name)
+            # it will be triggered when it gets garbage collected
+            self._finalizer = weakref.finalize(
```

Comment:
shouldn't this been done when the SnowflakeTable object is created? instead of at selectable